### DISCUSSION TITLE
Conditional attribute appearance works in React but doesn't in Next.js?

### DISCUSSION THREAD

**Discussion by Enkratia** [vercel/next.js#58590] [Help] [answered]
### Summary

Why `data-text-btn-active="false"` in html when value is false? Shouldn't attribute be ommited accordingly to https://github.com/facebook/react/blob/8b88ac2592c5f555f315f9440cbb665dd1e7457a/packages/react-dom/src/shared/DOMProperty.js  (line 30 && 167)?

This attribute is ommited In React but when i do the same in Next.js - result is different (attribute doesn`t disappear)? I do something wrong?

```
const TextButton: React.FC<TextButtonProps> = () => {
  return (
    <button data-text-btn-active={isActive ? true : false} ></button>
  );
};
```

**Comment by icyJoseph** [comment, marked as answer]
Not really right?

Type, `react.new` in your browser url and paste this:

```js
import "./styles.css";
import {renderToString} from 'react-dom/server'

const TextButton = ({isActive}) => {
  return (
    <button data-text-btn-active={isActive ? true : false} >hello</button>
  );
};

console.log(renderToString(<TextButton/>))

export default function App() {
  return (
    <div className="App">
      <TextButton/>
    </div>
  );
}

```

You'll see that the console and elements tabs, have `data-text-btn-active="false"`.

If we look at the code you linked, there's two functions we care about:

- shouldRemoveAttribute
- shouldRemoveAttributeWithWarning

In this case, for a boolean value, we enter this branch, within `shouldRemoveAttribute`:

```js
if (
    shouldRemoveAttributeWithWarning(
      name,
      value,
      propertyInfo,
      isCustomComponentTag,
    )
  )
```

So, `shouldRemoveAttributeWithWarning`, has to return true to remove this attribute.

For a boolean, we have this, within `shouldRemoveAttributeWithWarning`:

```js
    case 'boolean': {
      if (isCustomComponentTag) {
        return false;
      }
      if (propertyInfo !== null) {
        return !propertyInfo.acceptsBooleans;
      } else {
        const prefix = name.toLowerCase().slice(0, 5);
        return prefix !== 'data-' && prefix !== 'aria-';
      }
    }
```

Assuming we end up in:

```js
        const prefix = name.toLowerCase().slice(0, 5);
        return prefix !== 'data-' && prefix !== 'aria-';
```

We can rewrite it into a testable function:

```js
const testName = (name) => {
  const prefix = name.toLowerCase().slice(0, 5);
  return prefix !== 'data-' && prefix !== 'aria-';
};
```

And now we run, `testName("data-text-btn-active")`, and it returns `false`, so we should not remove with warning.

Then we can assume that ` if (propertyInfo !== null)` is not applying, for custom attributes, or aria labels, and so we return `false`, to not remove at all.

**Comment by Enkratia** [comment]
@icyJoseph, it works with native attributes like `disabled`, `hidden.` But doesn't with custom attribues like `data-*` and `aria-*`? Right?

**Reply by icyJoseph** [reply]
Yes, because those are kind of special, since their presence does convey meaning.

**Reply by Enkratia** [reply]
Thank you
